### PR TITLE
Adds Validation Rules

### PR DESCRIPTION
**Pull request by NeedleInAJayStack**
This adds 19 new validation rules from graphql-js as well as all associated tests. As bugs were exposed through this testing, relevant fixes were made. This should improve usability of GraphQL APIs built on this package, since APIs will now automatically return better error messages to clients.

Since this is such a large merge request, I've organized the commits so that each new rule is specified in a `feature` commit. Any required fixes or adjustments were made in `fix` commits before the `feature` commit in which the rule was added. The MR may be easier to review commit-by-commit.

### CODE REVIEW

File: Sources/GraphQL/Type/Definition.swift [hunk 1]
```diff
@@ -171,35 +172,22 @@ public final class GraphQLScalarType {
...
+    let parseLiteral: (Value) throws -> Map
 
     public init(
         name: String,
         description: String? = nil,
         serialize: @escaping (Any) throws -> Map,
-        parseValue: @escaping (Map) throws -> Map,
-        parseLiteral: @escaping (Value) throws -> Map
+        parseValue: ((Map) throws -> Map)? = nil,
+        parseLiteral: ((Value) throws -> Map)? = nil
```
Comment (NeedleInAJayStack) [inline Sources/GraphQL/Type/Definition.swift:183]: Note that the combination of these two initializer changes is that the Scalar public API is preserved.

**Comment by NeedleInAJayStack** [comment]
@paulofaria @d-exclaimation Hey guys, let me know if there's something I can do to make this easier to review. I realize that it's very big and I'd be willing to split it up into smaller PRs if that helps.

**Review by paulofaria — DISMISSED** [review]
I thought I had already approved this! 😅

**Comment by NeedleInAJayStack** [comment]
@paulofaria Thanks! In applying it to my big complex GraphQL schema, I found and fixed one small bug. Please re-approve if that one looks okay! https://github.com/GraphQLSwift/GraphQL/compare/33fa691e106911893265995f1be99efd965d6e1b..a4063a10941592c45c3dd92f3f0c5f9d70043aca

**Review by paulofaria — APPROVED** [review]

**Pull request merged**